pandas/tests/groupby/aggregate/test_aggregate.py

```diff
@@ -313,3 +313,79 @@ def test_order_aggregate_multiple_funcs():
     expected = pd.Index(['sum', 'max', 'mean', 'ohlc', 'min'])
 
     tm.assert_index_equal(result, expected)
+
+
+class TestKeywordAggregation:
+
+    def test_agg_relabel(self):
+        df = pd.DataFrame({"group": ['a', 'a', 'b', 'b'],
+                           "A": [0, 1, 2, 3],
+                           "B": [5, 6, 7, 8]})
+        result = df.groupby("group").agg(
+            a_max=("A", "max"),
+            b_max=("B", "max"),
+        )
+        expected = pd.DataFrame({"a_max": [1, 3], "b_max": [6, 8]},
+                                index=pd.Index(['a', 'b'], name='group'),
+                                columns=['a_max', 'b_max'])
+        tm.assert_frame_equal(result, expected)
+
+        # order invariance
+        result = df.groupby('group').agg(
+            b_min=("B", "min"),
+            a_min=("A", min),
+            a_max=("A", "max"),
+            b_max=("B", "max"),
+        )
+        expected = pd.DataFrame({"b_min": [5, 7],
+                                 "a_min": [0, 2],
+                                 "a_max": [1, 3],
+                                 "b_max": [6, 8]},
+                                index=pd.Index(['a', 'b'], name='group'),
+                                columns=['b_min', 'a_min', 'a_max', 'b_max'])
+        if not compat.PY36:
+            expected = expected[['a_max', 'a_min', 'b_max', 'b_min']]
+        tm.assert_frame_equal(result, expected)
+
+    def test_agg_relabel_non_identifier(self):
+        df = pd.DataFrame({"group": ['a', 'a', 'b', 'b'],
+                           "A": [0, 1, 2, 3],
+                           "B": [5, 6, 7, 8]})
+
+        result = df.groupby("group").agg(**{'my col': ('A', 'max')})
+        expected = pd.DataFrame({'my col': [1, 3]},
+                                index=pd.Index(['a', 'b'], name='group'))
+        tm.assert_frame_equal(result, expected)
+
+    def test_duplicate_raises(self):
+        # TODO: we currently raise on multiple lambdas. We could *maybe*
+        # update com.get_callable_name to append `_i` to each lambda.
+        df = pd.DataFrame({"A": [0, 0, 1, 1], "B": [1, 2, 3, 4]})
+        with pytest.raises(SpecificationError, match="Function names"):
+            df.groupby("A").agg(a=("A", "min"), b=("A", "min"))
```

Comment:
Why doesn't this work? These are not lambda functions. Do you mean that the ("selected colname", "aggfunc `__name__`") pairs must be all unique?

Reply:
I'd like to add, that ideally:
* the ("selected colname", "aggfunc `__name__`") pairs should **not** have to be all unique
* the ("renamed dest col", "selected colname", "aggfunc `__name__`") pairs **must** be all unique

If the above holds, then using multiple lambdas and/or multiple partials based on the same base function would work and this would solve the issues mentioned in the initial post of #18366.

I suppose that this implementation-related limitation is the one you were referring to at the end of your comment in https://github.com/pandas-dev/pandas/issues/18366#issuecomment-492784947

Reply:
> Do you mean that the ("selected colname", "aggfunc __name__") pairs must be all unique?

Yes. This could perhaps be relaxed in the future.